### ISSUE TITLE
DKIM looses DNS if computer sleeps and change network when restarting

### ISSUE THREAD

**Issue by sbsrouteur** [lieser/dkim_verifier#257]
Hello, 

I consistently have the issue that DKIM shows DNS timeout when my computer wakes up from sleep and has changed network location.
Stopping, and restarting TB "fixes" the issue, but if would be nice if DKIM could handle this.

I have tried various setting for the DNS but none seem to be able to work properly when the system wakes up after changing network.

My current setting is :
 - Javascript DNS library
 - Do not use OS dns servers
 - forced server name 8.8.8.8
 - No Proxy
 - Computer is Win10 Pro

Thank you

**Comment by lieser** [comment]
My first advise would have been to set a fixed DNS server, but you already did that.
Can you also observe this problem if you do not immediately view an e-mail after the sleep? Seems like it takes a while for the network connecting to be established after the sleep.

You could try out to change the hidden [`dns.jsdns.autoResetServerAlive`](https://github.com/lieser/dkim_verifier/wiki/DNS#advanced-options) setting to true.
See https://github.com/lieser/dkim_verifier/wiki/Debug#set-hidden-options for how to do that,

I would also recommend to try out the libunbound resolver, which has I think a better behavior with regards to unstable/changing network configurations.

**Comment by sbsrouteur** [comment]
Hello,

Thank you for the feed back. I have actuvated the autoResetServerAlive, and I see some changes in the way the DNS request are sent. But it seems, I got mistaken in understanding the error.

The error seem to occur when I am on VPN with my company. We use OpenVPN with forced DNS to our company DNS servers. This seems to be a problem for the javascript dns resolver. Would it be possible to have a copy of the libunbound dll in this repo for downloading, as the libunbound development has ceased and the download link is now broken?

Regards,

**Comment by lieser** [comment]
You could try out the dll in the zip posted in https://github.com/lieser/dkim_verifier/issues/170#issuecomment-595082202. Note that you need all three dlls, and some weird behavior can occurs if you move them, which can be avoided by explicitly mention all three dlls. See my comment https://github.com/lieser/dkim_verifier/issues/170#issuecomment-534318043 in the same issue.

As for providing the dlls in this repo:
I do not want to officially distribute them myself without also building them myself (or better in a CI) from source. Both to avoid any potential licensing issues and so I can be sure what I actually do distribute.
I know this may seem like only a small step to the links I already have (which unfortunately seem to be down now - hopefully something temporary), and the zip files attached to some comments in the issues. But it is still not something I am comfortable doing.

Building and distributin them myself is tracked in #122.

**Comment by sbsrouteur** [comment]
Hello,

I tried the libunbound dll from #170 and it seems to work, but I continue having error, I do not understand.

Exemple 1: DKIM works fine : 
![image](https://user-images.githubusercontent.com/11095871/117033470-ae417f80-ad02-11eb-852f-ac2586956a94.png)

ML rerouting makes DKIM red, but all get resolved properly.

Exemple 2 : DKIM crash ??
![image](https://user-images.githubusercontent.com/11095871/117033616-d4ffb600-ad02-11eb-8e44-791c8224f99c.png)
internal error (I can find some stack traces in the debug console)
(email sent by a colleague of mine threw our server, so email is supposed to be signed properly)

Exemple 3 : DKIM works
![image](https://user-images.githubusercontent.com/11095871/117034476-a209f200-ad03-11eb-8212-790a43fd36b9.png)

Same colleague sends an email via a mailing list, and all is signed ok.

Actually when looking for the above examples, I am under the impression that emails sent directly to me seem to trigger an error, while the one I get via a mailling list seem handled OK. Am I missing something ?

**Comment by lieser** [comment]
#### Exemple 1: `DKIM_POLICYERROR_WRONG_SDID` / `Mauvais signataire (devrait être $1)`

You should have a look at your configured sign rules. See also
- https://github.com/lieser/dkim_verifier/wiki/FAQ#the-add-on-incorrectly-says-that-en-e-mail-should-be-signed
- https://github.com/lieser/dkim_verifier/wiki/Sign-rules

#### Exemple 2: internal error

If possible, please send an affected email to lieser+dkim@posteo.net so I can investigate the problem.

**Comment by sbsrouteur** [comment]
Thanks for the help, 

I sent you an email to test (dkim seems ok) with a sample as attachment

**Comment by lieser** [comment]
Thanks for the email, I however can not reproduce the internal error. Please make sure you have the latest version (4.0,0) installed.

If you still have the error with the latest version, please enable debugging in the advanced option and post or send me the output from the error console https://github.com/lieser/dkim_verifier/wiki/Debug#view-error-and-debug-messages.

**Comment by sbsrouteur** [comment]
Hello, 

I am on 4.0.0 (update Apr 18th).

I clicked on one of the mail and get this in debug console : 
![image](https://user-images.githubusercontent.com/11095871/117141740-2d8c8d00-adaf-11eb-8365-a35e23f5cd49.png)

**Comment by lieser** [comment]
That `InvalidStateError: A mutation operation was attempted on a database that did not allow mutations` error looks strange. I did find https://bugzilla.mozilla.org/show_bug.cgi?id=1633209, but I guess you are not trying to debug the extension?

Anyway, please try the following:
1. disable all other extensions
2. close all tabs besides the main one
3. restart TB
4. try to reproduce the error

If that does not help:
- Can you change add-on settings? Or does the same error occur?
- Can you change sign rules, or edit/delete a stored key? Or does the same error occur?

To try to get out of that state, you could try the following. **Important:** This should result in TB deleting the storage of the add-on. Which means you will lose your settings of the add-on (including sign rules and stored keys). Migration of old setting should happen again.
1. Remove the add-on
2. Restart TB
3. Install the add-on

**Comment by sbsrouteur** [comment]
Hello,

I was indeed "debugging". I did place some breakpoints in the code to try look and see if anything would popup that might be helpfull.

I did uninstall, restart and reinstall. And some prior settings came back (ie the dns and display settings).
Right now, I do not see the issue, but I just restarted. 

I suggest to give it a go, and if nothing comes up then close this issue. I have been using your extension from quite a while now, and maybe some update/upgrade did leave some settings in an unexpected state, which the reinstall fixed.
